### PR TITLE
Np 522/create authorizer from authenticated approval dao

### PR DESCRIPTION
**Pull request by KolombetM**
https://nanopay.atlassian.net/browse/NP-522

AuthenticatedApprovalDAOAuthorizer was created to replace AuthenticatedApprovalDAO. Everything from nanopay/approval was moved to foam/approval

Approval was moved from nanopay to foam. So it is repeat of this pr:
https://github.com/nanoPayinc/NANOPAY/pull/9633

@jlhughes 
@tharmaman

### CODE REVIEW

File: src/foam/nanos/approval/ApprovalRequestNotificationDAO.java
Comment (jlhughes) [inline src/foam/nanos/approval/ApprovalRequestNotificationDAO.java]: The proper way to do this, is 
user.doNotify(...) 
rather than create a notification.  The users notification settings will handle sending an email, sms, in-app notification - whatever they have opted in to.

File: src/foam/nanos/approval/services.jrl
Comment (noodlemoodle) [inline src/foam/nanos/approval/services.jrl]: this one is true by default in easydao

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (noodlemoodle) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:35]: this should also check if the user has read permission, and throw exception if the user doesn't have read permission, and is also not the approver

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (noodlemoodle) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]: i think just the permission check is sufficient here

Comment (KolombetM) [reply src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]: Does it mean that user will be able to set himself as approver, but won't be able to approve request?

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (KolombetM) [reply src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:35]: If I get it correct, you want me to add check:
authService.check(x, "approval.read"). Then, who should have read permission for concrete request and how to grant it?

Comment (noodlemoodle) [reply src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:35]: for example, authorizeonread for an approvalrequest should return true for admins even if they are not the approver of the request

Comment (KolombetM) [reply src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:35]: Admin will be handled by checkGlobalRead

Comment (noodlemoodle) [reply src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:35]: okok thats fine then, i can't think of any other type of user authorized to read someone else's approve request

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (noodlemoodle) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]:
```suggestion
```

File: src/foam/nanos/approval/ApprovalRequestNotificationDAO.java
Comment (tharmaman) [inline src/foam/nanos/approval/ApprovalRequestNotificationDAO.java]: @noodlemoodle don't need approvable hash key for remove since its just the object itself, no check for properties or anything

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (tharmaman) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]: Shouldnt this be approval.update. + oldObj.getId(), since its not a global check. Automatically a user with approval.update.* will be able to update all approvals.

Might make more sense if authorizeOnRead, authorizeOnUpdate and authorizeOnDelete check if the user id is equal to the approver id rather than an explicit permission string check because then its either check if a user has wildcard and they'll see all approval requests or check if they have a permission to individually read that request itself which I don't think works best with approvals. I remember this specific check for if the user is the approver was done in the approvalDAO somewhere before @jlhughes @lchanmann @noodlemoodle

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (tharmaman) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]: See above.

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (tharmaman) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:64]: See above.

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (tharmaman) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:71]: See above.

**Review by tharmaman — CHANGES_REQUESTED** [review]
Regarding the authorizer

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (tharmaman) [reply src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]: I see ruby talked about it in read

File: src/foam/nanos/approval/ApprovalRequestNotificationDAO.java
Comment (noodlemoodle) [reply src/foam/nanos/approval/ApprovalRequestNotificationDAO.java]: but this is for removing the approvalrequest no? not operations.remove @tharmaman

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (noodlemoodle) [reply src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]: makes sense to me for read and update, but i'm not sure if users should be able to delete approvalrequests unless they have explicit permission

Comment (KolombetM) [reply src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]: So what should be done for delete?

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (KolombetM) [reply src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:64]: What is the reason to check if user is approver in globalRead? Doesn't it break the idea of globalRead permission?

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (tharmaman) [reply src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]: Ya so delete should be then just admin or the system:

```
foam.nanos.auth.User user = (foam.nanos.auth.User) x.get("user");
if ( user != null && ( user.getId() == foam.nanos.auth.User.SYSTEM_USER_ID || user.getGroup().equals("admin") || user.getGroup().equals("system") ) ) return;
throw new AuthorizationException("Approval requests can only be created by the system");
```

Is what is on the LiquidApprovalRequestAuthorizer

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (tharmaman) [reply src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:64]: Nvm ya leave it global i misread.

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (tharmaman) [reply src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:71]: Nvm leave it as the wildcard I misread.

File: src/foam/nanos/approval/ApprovalRequestNotificationDAO.java
Comment (tharmaman) [reply src/foam/nanos/approval/ApprovalRequestNotificationDAO.java]: My bad. But why does the Approvable_Hash_Key have to exist here? Isn't the object id sufficient? Obj_Id either refers to the Approvable or the object itself. @noodlemoodle

Comment (noodlemoodle) [reply src/foam/nanos/approval/ApprovalRequestNotificationDAO.java]: hmmmm make sense but I think for approvable the obj_id is just the daokey + id of obj being updated now, so there could be multiple approvalrequests for different update operations for the same "objId" right? so its not guaranteed the correct "fulfilled request" will be returned  @tharmaman

Comment (tharmaman) [reply src/foam/nanos/approval/ApprovalRequestNotificationDAO.java]: Yeah so in the rework i did, approvables now use GUID and the GUID is stored as the obj_id and a new property lookup_id is on approvable which has that hash. So I'll remove the Approvable HAsh Key in my PR because we don't even set it in the approvableAwareDAO @noodlemoodle

Comment (noodlemoodle) [reply src/foam/nanos/approval/ApprovalRequestNotificationDAO.java]: oh true ok makes sense i should look thru the rework first @tharmaman

File: src/files.js
Comment (lchanmann) [inline src/files.js]: For consistency, can we rename it to 'foam/nanos/approval/refinements'?

File: src/foam/nanos/approval/UserRefines.js
Comment (lchanmann) [inline src/foam/nanos/approval/UserRefines.js]: 'foam.nanos.auth.User.permission.viewApprovalRequests' permission is defined at nanopay side. We either leave this refinement at nanopay or move the permission over as well. @jlhughes @tharmaman

File: src/foam/nanos/approval/services.jrl
Comment (lchanmann) [inline src/foam/nanos/approval/services.jrl]: Please update "approvalRequestDAO" in src/services instead of a new file.

File: src/foam/nanos/approval/ApprovalRequestNotificationDAO.java
Comment (lchanmann) [inline src/foam/nanos/approval/ApprovalRequestNotificationDAO.java]: ComplianceApprovalRequest is from nanopay.

File: src/foam/nanos/approval/ApprovalRequestNotificationDAO.java
Comment (lchanmann) [inline src/foam/nanos/approval/ApprovalRequestNotificationDAO.java]: If we let approval request generates the notification, the subclass eg. compliance approval request can override the implementation as needed.

File: src/foam/nanos/approval/ApprovalRequestNotificationDAO.java
Comment (lchanmann) [inline src/foam/nanos/approval/ApprovalRequestNotificationDAO.java]: should canceled status be added?

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (lchanmann) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]: Please optimize java imports.

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (lchanmann) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]: userId is unused?

File: src/foam/nanos/approval/test/ApprovalStatusTest.java
Comment (lchanmann) [inline src/foam/nanos/approval/test/ApprovalStatusTest.java]: optimize imports

File: src/foam/nanos/approval/test/ApprovalStatusTestExecutor.java
Comment (lchanmann) [inline src/foam/nanos/approval/test/ApprovalStatusTestExecutor.java]: import from nanopay

File: src/foam/nanos/approval/test/ApprovalTestExecutor.java
Comment (lchanmann) [inline src/foam/nanos/approval/test/ApprovalTestExecutor.java]: imports from nanopay

File: src/foam/nanos/approval/test/ApprovalTestExecutor.java
Comment (lchanmann) [inline src/foam/nanos/approval/test/ApprovalTestExecutor.java]: diff detected some changes made. @tharmaman can you verify if it's fine?

File: src/foam/nanos/approval/test/LiquidApprovalCreateRejectTest.java
Comment (lchanmann) [inline src/foam/nanos/approval/test/LiquidApprovalCreateRejectTest.java]: should liquid specific test be on foam?

File: src/foam/nanos/approval/test/LiquidApprovalDeleteRejectTest.java
Comment (lchanmann) [inline src/foam/nanos/approval/test/LiquidApprovalDeleteRejectTest.java]: same as above

File: src/foam/nanos/approval/test/LiquidApprovalTest.java
Comment (lchanmann) [inline src/foam/nanos/approval/test/LiquidApprovalTest.java]: same as above

File: src/foam/nanos/approval/test/LiquidApprovalUpdateRejectTest.java
Comment (lchanmann) [inline src/foam/nanos/approval/test/LiquidApprovalUpdateRejectTest.java]: same as above

**Review by lchanmann — CHANGES_REQUESTED** [review]

File: src/foam/nanos/approval/test/LiquidApprovalCreateRejectTest.java
Comment (Mykola97) [reply src/foam/nanos/approval/test/LiquidApprovalCreateRejectTest.java:10]: It was there all the time. Where do you want me to move it?

File: src/foam/nanos/approval/refinements.js
Comment (lchanmann) [inline src/foam/nanos/approval/refinements.js]:
```suggestion
    name: 'UserRefinement',
```
The class name should be capitalized.

**Review by KolombetM — COMMENTED** [review]
I made a mistake and moved nanopay specific files to foam. 

- All nanopay specific files was left in nanopay.
- foam/src/services file was changed instead of nanopay/approval/service.jrl

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (lchanmann) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]: unused imports

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (lchanmann) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]:
```suggestion
        User user = ((Subject) x.get("subject")).getRealUser();
```
And replace userId with user.getId() on line 48.

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (lchanmann) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]:
```suggestion
        User user = ((Subject) x.get("subject")).getRealUser();
```

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (lchanmann) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js]:
```suggestion
        User user = ((Subject) x.get("subject")).getRealUser();;
```

File: src/foam/u2/md/OverlayDropdown.js
Comment (lchanmann) [inline src/foam/u2/md/OverlayDropdown.js]: please remove debugger

**Review by lchanmann — APPROVED** [review]

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (noodlemoodle) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:18]:
```suggestion
```

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (noodlemoodle) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:17]:
```suggestion
    'foam.util.SafetyUtil'
```

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (noodlemoodle) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:46]:
```suggestion
        if ( user == null || ! SafetyUtil.equals(approvalRequest.getApprover(), user.getId()) && ! ( user.getId() == foam.nanos.auth.User.SYSTEM_USER_ID || user.getGroup().equals("admin") || user.getGroup().equals("system") ) ) {
```

File: src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js
Comment (noodlemoodle) [inline src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js:55]:
```suggestion
        if ( user == null  || ! ( user.getId() == foam.nanos.auth.User.SYSTEM_USER_ID || user.getGroup().equals("admin") || user.getGroup().equals("system") ) ) {
```

**Review by noodlemoodle — APPROVED** [review]
some spacing suggestions otherwise lgtm

File: src/foam/nanos/approval/UserRefines.js
Comment (jlhughes) [reply src/foam/nanos/approval/UserRefines.js]: @Mykola97 @lchanmann Please move over.

**Pull request merged**
